xbmc/listproviders/MultiProvider.h

```diff
@@ -0,0 +1,51 @@
+/*
+ *      Copyright (C) 2013 Team XBMC
+ *      http://xbmc.org
+ *
+ *  This Program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2, or (at your option)
+ *  any later version.
+ *
+ *  This Program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with XBMC; see the file COPYING.  If not, see
+ *  <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#pragma once
+
+#include <vector>
+#include <map>
+#include "IListProvider.h"
+#include "threads/CriticalSection.h"
+
+typedef std::shared_ptr<IListProvider> IListProviderPtr;
+
+/*!
+ \ingroup listproviders
+ \brief A listprovider that handles multiple individual providers.
+ */
+class CMultiProvider : public IListProvider
+{
+public:
+  CMultiProvider(const TiXmlNode *first, int parentID);
+  
+  virtual bool Update(bool forceRefresh) override;
+  virtual void Fetch(std::vector<CGUIListItemPtr> &items) const override;
+  virtual bool IsUpdating() const override;
+  virtual void Reset(bool immediately = false) override;
+  virtual bool OnClick(const CGUIListItemPtr &item) override;
+  virtual bool OnInfo(const CGUIListItemPtr &item) override;
+  virtual bool OnContextMenu(const CGUIListItemPtr &item) override;
+  
+protected:
+  std::vector<IListProviderPtr> m_providers;
+  mutable std::map<CGUIListItem*, IListProviderPtr> m_itemMap; 
```

Comment:
The idea is that the MultiProvider needs to know which sub-provider will be sent Click/Info/Context messages so it needs to keep track of which content element belong to which sub-provider.

The only way I see to maintain that mapping is to build it during Fetch().  Unfortunately, Fetch is const in the base class.

I could maybe change that without too much impact on other code.  Would that be preferable to a mutable here?